launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/LaunchBarManager.java

```diff
@@ -475,8 +489,14 @@ public void setActiveLaunchDescriptor(ILaunchDescriptor descriptor) throws CoreE
 			Activator.trace("resync for " + descriptor); //$NON-NLS-1$
 			return;
 		}
-		if (descriptor != null && !descriptors.containsValue(descriptor)) {
-			throw new IllegalStateException(Messages.LaunchBarManager_1);
+		if (descriptor != null) {
+			boolean isContained;
+			synchronized (descriptors) {
+				isContained = descriptors.containsValue(descriptor);
+			}
+			if (!isContained) {
+				throw new IllegalStateException(Messages.LaunchBarManager_1);
```

Comment:
I am pretty sure you don't need to save off the `isContained` here, I think throwing from within the synchronized block is fine:

```java
syncrhonized(descriptors) {
    if (descriptor != null && !descriptors.containsValue(descriptor)) {
        throw new IllegalStateException(Messages.LaunchBarManager_1);
    }
}
```

If you need to optimize to not get the lock if `descriptor == null` then a partial split is fine too:

```java
if (descriptor != null) {
    syncrhonized(descriptors) {
        if (!descriptors.containsValue(descriptor)) {
            throw new IllegalStateException(Messages.LaunchBarManager_1);
        }
    }
}
```